### PR TITLE
Add traverseWithKey_ for HashMaps

### PR DESCRIPTION
**Pull request by waldheinz**
I needed this, and if you think it's ok I'll clean it up and add tests.

### CODE REVIEW

File: Data/HashMap/Base.hs [hunk 1]
```diff
@@ -768,6 +769,19 @@ traverseWithKey f = go
...
 {-# INLINE traverseWithKey #-}
 
+-- | /O(n)/ Map each element of a map to an action, evaluate these
+-- actions, and ignore the results.
+traverseWithKey_ :: Applicative f => (k -> v1 -> f b) -> HashMap k v1
+                -> f ()
+traverseWithKey_ f = go
+  where
+    go Empty                 = pure ()
+    go (Leaf _ (L k v))      = f k v *> pure ()
```
Comment (tibbe) [inline Data/HashMap/Base.hs]: Check the Travis build failure. `*>` is not in scope. Make sure this change works with the latest 3 major GHC releases.

**Comment by tibbe** [comment]
One minor comment and please add a test. Thanks!

**Comment by waldheinz** [comment]
I added a test which uses `State` to re-build the same `Map` we're traversing and then checks if the result of this process equals the original map. Is this sufficient?

**Comment by waldheinz** [comment]
PS: For the test I added a dependency on `mtl`. The `mtl` was an indirect dependency for the tests already, so this is hopefully acceptable. But I don't know what to use for the version lower bound.

**Comment by tibbe** [comment]
Test looks fine. Also look into my comment above. `*>` isn't imported correctly on all GHC versions.

**Comment by waldheinz** [comment]
I already updated my branch for the missing `*>` import, that's why the tests are passing now. :-)

File: tests/HashMapProperties.hs [hunk 2]
```diff
@@ -6,6 +6,7 @@
 module Main (main) where
 
 import Control.Monad ( guard )
+import Control.Monad.State ( execState, get, put )
```
Comment (phadej) [inline tests/HashMapProperties.hs]: wont `transformers` (i.e. `Control.Monad.Trans.State`) be enough? (though it's in tests, so probably not a big deal)

**Comment by treeowl** [comment]
It seems a bit weird to have this but not to have `foldMapWithKey`.

Like most of the rest of the library, this function does not document anything about the order of the traversal, or even whether that's determined by the set of keys in the map (my guess is no). I think more documentation is in order!

**Comment by sjakobi** [comment]
FWIW, I'm in favour of the addition of `traverseWithKey_`. Are you planning to work on this further @waldheinz?

Since this PR needs at least a rebase, I'll mark it as a draft for now.

**Comment by sjakobi** [comment]
> FWIW, I'm in favour of the addition of `traverseWithKey_`.

I just noticed that `containers` doesn't provide `traverseWithKey_` though. Any idea why @treeowl?

**Comment by waldheinz** [comment]
I just played around with a rebase but it seems after 5 years my Haskell foo is a bit weak to come up with something quick. If there is interest I'm willing to work on this next week. But please dicuss if this addition is actually welcome first so effort is not wasted.

**Comment by treeowl** [comment]
> > FWIW, I'm in favour of the addition of `traverseWithKey_`.
> 
> I just noticed that `containers` doesn't provide `traverseWithKey_` though. Any idea why @treeowl?

Probably just that `foldrWithKey` is very powerful. I'm generally wary of new functions that are simple applications of the existing API.

**Comment by sjakobi** [comment]
> > > FWIW, I'm in favour of the addition of `traverseWithKey_`.
> > 
> > 
> > I just noticed that `containers` doesn't provide `traverseWithKey_` though. Any idea why @treeowl?
> 
> Probably just that `foldrWithKey` is very powerful. I'm generally wary of new functions that are simple applications of the existing API.

Good point! Is `foldrWithKey` sufficient for your usecase, @waldheinz?

I think we should at least add an example of how to define `traverseWithKey_` to the documentation for `foldrWithKey`.

I've also wondered whether [`bitraverse_`](http://hackage.haskell.org/package/base-4.14.0.0/docs/Data-Bifoldable.html#v:bitraverse_) could be useful in this context, but the type is a bit different.

---

If none of this works, I'd suggest to send a proposal regarding this addition to the [libraries mailing list](https://mail.haskell.org/cgi-bin/mailman/listinfo/libraries).

**Comment by sjakobi** [comment]
@waldheinz: https://github.com/haskell-unordered-containers/unordered-containers/pull/297 documents how to define `traverseWithKey_` in terms of `foldrWithKey`.

**Comment by sjakobi** [comment]
Closing because this PR has stalled. The latest discussion on this addition is in #297.

**Comment by sjakobi** [comment]
My bad. In https://github.com/haskell-unordered-containers/unordered-containers/pull/297 @treeowl and I actually agreed that we want this API addition.

@waldheinz Are you willing to rebase this PR and drive it to completion? I promise a speedy review.

**Pull request closed without merging**